### PR TITLE
[Perf]Boogu-Image Skip dense attention masks

### LINKED ISSUE #6862: [Performance]: Boogu-Image passes an all-ones attention mask, forcing a device sync per attention call
#### Summary

Boogu-Image builds an attention mask on every forward, unconditionally. At batch size 1 that mask is all-ones, but the FlashAttention backend still has to *check* whether it is — and that check is a device-to-host copy followed by a stream synchronise. One per attention call, **50 per transformer forward, ~2800 per 1024x1024 request**.

The GPU work of the check is negligible (~250 us per forward). The cost is the pipeline drain it forces.We tested the Boogu-Image inference performance on one H200 and gathered the nsys timeline. It shows GPU idle before each FlashAttention kernel. 

Image

## Root cause

**1. The dense/varlen dispatch asks the GPU a question the host already knows.**
`vllm_omni/diffusion/attention/backends/flash_attn.py:417`:

```python
if attention_mask is not None and torch.any(~attention_mask):
    return self._forward_varlen_masked(query, key, value, attention_mask)
```

`torch.any(...)` returns a 0-dim CUDA tensor. Using it in a Python `if` calls `Tensor.__bool__`, which issues a `cudaMemcpyAsync` D2H plus a `cudaStreamSynchronize`. The 1-byte transfer is trivial; the ordering guarantee. 
it requires is not — the stream must drain everything queued before it, and the CPU's run-ahead is reset, so the GPU then waits for Python to reach the next launch. This runs on *every* call, whichever branch is finally taken.

**2. Boogu builds a mask unconditionally.**
`vllm_omni/diffusion/models/boogu_image/boogu_image_transformer.py:1277-1286`:

```python
joint_attention_mask = hidden_states.new_zeros(batch_size, max_seq_len, dtype=torch.bool)
for i, seq_len in enumerate(seq_lengths):
    joint_attention_mask[i, :seq_len] = True
```

At batch size 1, `seq_len == max_seq_len`, so the mask is all-ones and the answer
to the question in (1) is always `False`.

#### Proposed Fix

At the model side, pass `None` when the mask is dense.**

```python
# boogu_image_transformer.py, at each of the four mask sites
if all(s == max_seq_len for s in seq_lengths):
    joint_attention_mask = None          # dense: let the backend take the fast path
else:
    joint_attention_mask = hidden_states.new_zeros(...)
    ...
```
`seq_lengths` is already a Python list, so the predicate costs nothing. `BooguImageSelfAttention.forward` already builds `AttentionMetadata` only when the mask is not `None`, so the dense path at `flash_attn.py:426` takes over the execution path. 

### Report of performance regression

Not a regression. This cost has been present since Boogu-Image landed in #4995 — no change introduced it, and no benchmark got slower. 

### Misc discussion on performance

Several merged PRs address the same class of problem:

- #3763 — `[Perf][Wan2.2] Skip attention mask for zero-padded SP sequences to avoid varlen path`
- #4293 — `[Perf][qwen-image-edit] Skip attention-mask to avoid varlen path`
- #4645 — `[Perf][Flux2][HunyuanVideo1.5] Skip attention-mask to avoid varlen path`
- #4333 — `[HunyuanImage][Perf] opt prepare_attention_mask for e2e latency 6% reduction`
- #6542 — `[perf] Add mask-free TRTLLM packed-padding path`

All of them gate the mask on sequence parallelism:

```python
if (... and parallel_config.sequence_parallel_size > 1
        and parallel_config.mask_sp_padding
        and ctx.sp_padding_size > 0):
```

Boogu's mask is not an SP-padding mask — it is built unconditionally, so the cost is paid on the **default single-GPU path**, where every one of those gates is false. Boogu fell through all four cases. 

### Your current environment (if you think it is necessary)

```text
The output of `python collect_env.py`
```

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by AbelSara**
## Purpose

Boogu-Image unconditionally builds `joint_attention_mask` and `img_attention_mask`, including for a single request or an equal-length batch. In those cases the masks contain only `True` values and do not mask any tokens.

Passing an all-true CUDA mask to the FlashAttention backend still evaluates `torch.any(~attention_mask)` to select the dense or masked varlen path. Using the resulting CUDA scalar in a Python condition introduces a device-to-host synchronization for every attention call and repeatedly drains the GPU stream.

This change adds a shared `_make_attention_mask` helper that:

- returns `None` for single-sample and equal-length batches, allowing them to enter the dense FlashAttention path directly;
- preserves the original boolean padding mask for variable-length batches.

Both the joint and image attention-mask construction paths use the helper. This is mathematically equivalent for dense batches because an all-true mask does not exclude any token.

Fixes #6862 

## Test Plan
```bash
cd tests
pytest -s -v diffusion/models/boogu_image/test_boogu_image_transformer.py -m "core_model and cpu"
```

**server**
```bash
vllm serve /root/autodl-tmp/models/Boogu-Image-0.1-Base   --omni   --vae-use-slicing  --port 8091  --vae-use-tiling
```
**client**
```bash
curl -s http://localhost:8091/v1/images/generations -H "Content-Type: application/json" -d '{"model":"/root/autodl-tmp/models/Boogu-Image-0.1-Base","prompt":"A mountain lake at sunset, photorealistic, cinematic lighting","size":"512x512","num_inference_steps":50,"guidance_scale":4.0,"seed":42}' | jq -r '.data[0].b64_json' | base64 -d > output.png
```

### Unit Test
image

### E2E Test
**main branch**
image

**current branch**
image

**image result**
image

**speed up**：1.08x

### Kernel Line
**main branch**
image

image

**current branch**
image

**comparison**
image

### CODE REVIEW

**Comment by AbelSara** [comment]
Equal-length batches now skip the unnecessary attention mask, while variable-length batches still build the correct padding mask.

**Comment by hsliuustc0106** [comment]
add the A/B test result for e2e and attn ops

**Comment by AbelSara** [comment]
> add the A/B test result for e2e and attn ops

@hsliuustc0106 @Gaohan123 @Bounty-hunter 
Hi, this pr is ready for review now.

File: vllm_omni/diffusion/models/boogu_image/boogu_image_transformer.py [hunk 1]
```diff
@@ -50,6 +50,29 @@ def _join_prefix(prefix: str, name: str) -> str:
     return f"{prefix}.{name}" if prefix else name
 
 
+def _make_attention_mask(
```
Comment (Gaohan123) [inline vllm_omni/diffusion/models/boogu_image/boogu_image_transformer.py]: I suggest we can move it into utils as a sharing function which can be used by other models

Comment (AbelSara) [reply vllm_omni/diffusion/models/boogu_image/boogu_image_transformer.py]: ok, I'll mv it

**Comment by hsliuustc0106** [comment]
can you help check whether there are other redundant intermediate tensors or operations

**Comment by AbelSara** [comment]
> can you help check whether there are other redundant intermediate tensors or operations

Sure,  I’d be happy to do that.

**Review by Gaohan123 — APPROVED** [review]
LGTM. Thanks
